docs: Revamp README to be simpler (#761)
TorchJD/torchjd: README.md

<div align="center">
  <img src="docs/source/_static/logo-light-mode.png" alt="TorchJD" width="400"/>

> </div>

Reviewer: The previous version had cases to use the right logo if the user uses the dark theme or light theme. Is it possible to still have this?

---

> [![Doc](https://img.shields.io/badge/Doc-torchjd.org-blue?logo=data%3Aimage%2Fsvg%2Bxml%3Bbase64%2CPDI5NGJXd2dkbVZ5YzJsdmJqMGlNUzR3SWlCbGJtTnZaR2x1WnowaVZWUkdMVGdpSUhOMFlXNWtZV3h2Ym1VOUltNXZJajglMkJDajtoUjNKbFlYUmxaQ0IxYzJsdVp5QkxjbWwwWVRvZ2FIUjBjRG92TDJ0eWFYUmhMbTl5WndBdExUNEtDand4TFMwZ1EzSmxZWFJsWkNCMWMybHVaeUJLY21sallUb2daWE1nYUhSMGNEb3ZMMnh2WTJGc2FRQXRMVDRLT0NBZ0lHZHBkSFJvUFNJeU1EUTNMamN5Y0hRaUNpQWdJR2hsYVdkb2REMGlNakEwTnk0M01uQjBJZ29nSUNCSllXd2dkR2hsUFNJeU1EUTNMamN5SURJd05EY3VNamtoQ2lBZ0lIWnBaWGRDYjNnOUlqQWdNQ0F5TURRdU55Y2dNakEwTnk0eU1TQXlNakF3TURBdElERXVNQ0F3SURNME5pNHlNVE1nTkRZdU9ESXpJREF3UXpVd01EQWdOVFl3TURBaUNnPT0=)](https://torchjd.org)

Reviewer: Looks like this change made the logo TorchJD disappear for the doc badge.

I like the idea of making it a little smaller though, and 400 looks good on my screen, but possibly, we want to have a proportion of the full window if possible.

[![Static Badge](https://img.shields.io/badge/%F0%9F%92%AC_ChatBot-chat.torchjd.org-blue?logo=%F0%9F%92%AC)](https://chat.torchjd.org)
[![Tests](https://github.com/SimplexLab/TorchJD/actions/workflows/checks.yml/badge.svg)](https://github.com/SimplexLab/TorchJD/actions/workflows/checks.yml)
[![codecov](https://codecov.io/gh/SimplexLab/TorchJD/graph/badge.svg?token=8AUCZE76QH)](https://codecov.io/gh/SimplexLab/TorchJD)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/torchjd)](https://pypi.org/project/torchjd/)
[![Static Badge](https://img.shields.io/badge/PyTorch-%3E%3D2.3-blue?logo=pytorch&logoColor=white)](https://pytorch.org/)
[![Static Badge](https://img.shields.io/badge/Discord%20-%20community%20-%20%235865F2?logo=discord&logoColor=%23FFFFFF&label=Discord)](https://discord.gg/76KkRnb3nk)

**TorchJD** is a PyTorch library for training neural networks with **multiple losses**. It supports two complementary approaches:

> - **Scalarization** — combine losses into a single scalar before backprop, using methods from the literature (geometric mean, softmax weighting, etc.)

Reviewer: Let's avoid M and N dashes, they sound like LLMs.

```suggestion
- **Scalarization**: combine losses into a single scalar before backprop, using methods from the literature (geometric mean, softmax weighting, etc.)
```

> - **Jacobian descent** — compute the full Jacobian matrix and aggregate it into a conflict-aware update direction using state-of-the-art aggregators (UPGrad, MGDA, CAGrad, and many more)

Reviewer: ```suggestion
- **[Jacobian descent](https://arxiv.org/pdf/2406.16232)**: compute the Jacobian matrix of losses with respect to parameters and aggregate it into an update direction using state-of-the-art aggregators (UPGrad, MGDA, CAGrad, and many more). This in particular allows taking conflict free optimization directions.
```

With Jacobian descent, $g_1$ and $g_2$ are computed individually and carefully aggregated using an
aggregator $\mathcal A$. In this example, the aggregator is the Unconflicting Projection of
Gradients $\mathcal A_{\text{UPGrad}}$: it
projects each gradient onto the dual cone, and averages the projections. This ensures that the
update will always be beneficial to each individual objective (given a sufficiently small step
size). In addition to $\mathcal A_{\text{UPGrad}}$, TorchJD supports
[more than 10 aggregators from the literature](https://torchjd.org/stable/docs/aggregation).
**TorchJD** is a PyTorch library for training neural networks with **multiple losses**. It supports two complementary approaches:

- **Scalarization** — combine losses into a single scalar before backprop, using methods from the literature (geometric mean, softmax weighting, etc.)
- **Jacobian descent** — compute the full Jacobian matrix and aggregate it into a conflict-aware update direction using state-of-the-art aggregators (UPGrad, MGDA, CAGrad, and many more)

The full documentation is available at [torchjd.org](https://torchjd.org).

## Installation

> <!-- start installation -->

Reviewer: This was used to include a section of the README in the installation page of the documentation, don't we do this anymore? @ValerianRey

Reviewer 2: Yes, this is still needed. Let's re-add it please.

> additional dependencies — refer to the [installation docs](https://torchjd.org/stable/installation).

Reviewer: Avoid M dashes

Reviewer: I liked this paragraph better before.

> Scalarization methods combine losses into a single scalar loss, which is then optimized with standard gradient descent. This is the simplest approach and is often a strong baseline.

Reviewer: ```suggestion
Scalarization methods combine losses into a single scalar loss, which is then optimized with standard gradient descent. This is the simplest approach and is a good baseline.
```

> Jacobian descent computes the per-task gradients individually and aggregates them into a single conflict-aware update direction. This avoids the issue where averaging conflicting gradients harms one of the objectives.

Reviewer: avoiding conflict is rather specific to the upgrad aggregator. not sure how to formulate that.

from torch.nn import Linear, MSELoss, ReLU, Sequential
from torch.optim import SGD

from torchjd.autojac import mtl_backward, jac_to_grad
from torchjd.aggregation import UPGrad

shared = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU())
task1_head = Linear(3, 1)
task2_head = Linear(3, 1)
params = [*shared.parameters(), *task1_head.parameters(), *task2_head.parameters()]

optimizer = SGD(params, lr=0.1)
criterion = MSELoss()
aggregator = UPGrad()

inputs = torch.randn(16, 10)
features = shared(inputs)
loss1 = criterion(task1_head(features), torch.randn(16, 1))
loss2 = criterion(task2_head(features), torch.randn(16, 1))

mtl_backward([loss1, loss2], features=features)
jac_to_grad(shared.parameters(), aggregator)
optimizer.step()
optimizer.zero_grad()
```

> More usage examples — including the memory-efficient `autogram` engine, instance-wise risk minimization, and partial Jacobian descent — can be found [in the docs](https://torchjd.org/stable/examples/).

Reviewer: avoid M dashes.

| [Random](https://torchjd.org/stable/docs/aggregation/random#torchjd.aggregation.Random) | [RandomWeighting](https://torchjd.org/stable/docs/aggregation/random#torchjd.aggregation.RandomWeighting) | [Reasonable Effectiveness of Random Weighting: A Litmus Test for Multi-Task Learning](https://arxiv.org/pdf/2111.10603) |
| - | [SDMGradWeighting](https://torchjd.org/stable/docs/aggregation/sdmgrad#torchjd.aggregation.SDMGradWeighting) | [Direction-oriented Multi-objective Learning](https://arxiv.org/pdf/2305.18409) |
| [Sum](https://torchjd.org/stable/docs/aggregation/sum#torchjd.aggregation.Sum) | [SumWeighting](https://torchjd.org/stable/docs/aggregation/sum#torchjd.aggregation.SumWeighting) | - |
| [Trimmed Mean](https://torchjd.org/stable/docs/aggregation/trimmed_mean#torchjd.aggregation.TrimmedMean) | - | [Byzantine-Robust Distributed Learning: Towards Optimal Statistical Rates](https://proceedings.mlr.press/v80/yin18a/yin18a.pdf) |

## Release Methodology

> TorchJD follows [semantic versioning](https://semver.org/). Since the library is still in beta (`0.x.y`), we sometimes make interface changes in minor versions. Breaking changes are always documented in the [changelog](CHANGELOG.md) with migration instructions.

Reviewer: I think I prefered the previous version.

## Contribution

Please read the [Contributing guide](CONTRIBUTING.md).

Thanks to our amazing contributors:

> [![Contributors](https://stg.contrib.rocks/image?repo=SimplexLab/TorchJD&max=240&columns=18)](https://github.com/SimplexLab/TorchJD/graphs/contributors)

Reviewer: Also preferred previous version.

## Citation

> 

Reviewer: ```suggestion
If you use TorchJD for your research, please cite:
```

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="docs/source/_static/logo-dark-mode.png">
  <source media="(prefers-color-scheme: light)" srcset="docs/source/_static/logo-light-mode.png">

>   <img alt="Fallback image description" src="docs/source/_static/logo-light-mode.png" width="400">

Reviewer: Is there a way to center it?

Author: yeah there is, I can do that

[installation documentation](https://torchjd.org/stable/installation) for them.

## Usage

### Scalarization

Scalarization methods combine losses into a single scalar before backprop. This is the simplest
approach and is often a strong baseline.

```python
import torch
from torch.nn import Linear, MSELoss, ReLU, Sequential
from torch.optim import SGD

from torchjd.scalarization import GeometricMean

model = Sequential(Linear(10, 5), ReLU(), Linear(5, 1))
optimizer = SGD(model.parameters(), lr=0.1)
criterion = MSELoss()
scalarizer = GeometricMean()

inputs = torch.randn(16, 10)
task1_targets, task2_targets = torch.randn(16, 1), torch.randn(16, 1)

output = model(inputs)
losses = torch.stack([criterion(output, task1_targets), criterion(output, task2_targets)])
loss = scalarizer(losses)
loss.backward()
optimizer.step()
optimizer.zero_grad()
```

### Jacobian descent

Jacobian descent computes per-loss gradients individually and aggregates them into a single update
direction. Some aggregators, like [UPGrad](https://torchjd.org/stable/docs/aggregation/upgrad/),
are specifically designed to find conflict-free directions that are beneficial to all losses

> simultaneously.

Reviewer: nice

[![Static Badge](https://img.shields.io/badge/PyTorch-%3E%3D2.3-blue?logo=pytorch&logoColor=white)](https://pytorch.org/)
[![Static Badge](https://img.shields.io/badge/Discord%20-%20community%20-%20%235865F2?logo=discord&logoColor=%23FFFFFF&label=Discord)](https://discord.gg/76KkRnb3nk)

TorchJD is a PyTorch library for training neural networks with **multiple losses**. It supports
two complementary approaches:

- **Scalarization**: combine losses into a single scalar before backprop, using methods from the

>   literature (geometric mean, softmax weighting, etc.). This is often a good baseline.

Reviewer 2: ```suggestion
  literature (geometric mean, softmax weighting, [etc.](#supported-scalarizers)). This is often a good baseline.
```

Please test that this local link works.

- **[Jacobian descent](https://arxiv.org/pdf/2406.16232)**: compute the Jacobian matrix of losses
  with respect to parameters and aggregate it into an update direction using state-of-the-art

>   aggregators (UPGrad, MGDA, CAGrad, and many more). This in particular allows taking conflict-free

Reviewer 2: ```suggestion
  aggregators (UPGrad, MGDA, CAGrad, [and many more]()#supported-aggregators-and-weightings). 
  This in particular allows taking conflict-free
```

pip install "torchjd[quadprog_projector]"
```
<!-- end installation -->

This includes the dependencies required by UPGrad and DualProj. Some other aggregators may have
additional dependencies. Please refer to the
[installation documentation](https://torchjd.org/stable/installation) for them.

## Usage

### Scalarization

Scalarization methods combine losses into a single scalar before backprop. Here is how to change
a standard training loop to use scalarization:

```diff
  import torch
  from torch.nn import Linear, MSELoss, ReLU, Sequential
  from torch.optim import SGD

+ from torchjd.scalarization import GeometricMean

  model = Sequential(Linear(10, 5), ReLU(), Linear(5, 1))
  optimizer = SGD(model.parameters(), lr=0.1)
  criterion = MSELoss()
+ scalarizer = GeometricMean()

  inputs = torch.randn(16, 10)
  task1_targets, task2_targets = torch.randn(16, 1), torch.randn(16, 1)

  output = model(inputs)
- loss = criterion(output, task1_targets) + criterion(output, task2_targets)
- loss.backward()
+ losses = torch.stack([criterion(output, task1_targets), criterion(output, task2_targets)])
+ loss = scalarizer(losses)
+ loss.backward()
  optimizer.step()
  optimizer.zero_grad()

> ```

Reviewer 2: Is it possible to use the same examples as for JD, i.e. have a training loop with 8 batches of data? I would be nice if everything is the same between the two examples except for what matters: scalarization vs JD.

[![Static Badge](https://img.shields.io/badge/PyTorch-%3E%3D2.3-blue?logo=pytorch&logoColor=white)](https://pytorch.org/)
[![Static Badge](https://img.shields.io/badge/Discord%20-%20community%20-%20%235865F2?logo=discord&logoColor=%23FFFFFF&label=Discord)](https://discord.gg/76KkRnb3nk)

TorchJD is a PyTorch library for training neural networks with **multiple losses**. It supports
two complementary approaches:

- **Scalarization**: combine losses into a single scalar before backprop, using methods from the
  literature (geometric mean, softmax weighting, [etc.](#supported-scalarizers)). This is often a good baseline.
- **[Jacobian descent](https://arxiv.org/pdf/2406.16232)**: compute the Jacobian matrix of losses
  with respect to parameters and aggregate it into an update direction using state-of-the-art

>   aggregators (UPGrad, MGDA, CAGrad, [and many more]()#supported-aggregators-and-weightings). 

Reviewer 2: ```suggestion
  aggregators (UPGrad, MGDA, CAGrad, [and many more](#supported-aggregators-and-weightings)). 
```

  shared_module = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU())
  task1_module = Linear(3, 1)
  task2_module = Linear(3, 1)
  params = [
      *shared_module.parameters(),
      *task1_module.parameters(),
      *task2_module.parameters(),
  ]
  params = [*shared_module.parameters(), *task1_module.parameters(), *task2_module.parameters()]

  loss_fn = MSELoss()
  optimizer = SGD(params, lr=0.1)
+ aggregator = UPGrad()

  inputs = torch.randn(8, 16, 10)  # 8 batches of 16 random input vectors of length 10
  task1_targets = torch.randn(8, 16, 1)  # 8 batches of 16 targets for the first task
  task2_targets = torch.randn(8, 16, 1)  # 8 batches of 16 targets for the second task

> 

Reviewer: I think the example doesn't work anymore because those are not defined.

  shared_module = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU())
  task1_module = Linear(3, 1)
  task2_module = Linear(3, 1)
  params = [*shared_module.parameters(), *task1_module.parameters(), *task2_module.parameters()]

  loss_fn = MSELoss()
  optimizer = SGD(params, lr=0.1)
+ aggregator = UPGrad()

  inputs = torch.randn(8, 16, 10)
  task1_targets = torch.randn(8, 16, 1)

>   task2_targets = torch.randn(8, 16, 1)

Reviewer 2: ```suggestion
  inputs = torch.randn(8, 16, 10)  # 8 batches of 16 random input vectors of length 10
  task1_targets = torch.randn(8, 16, 1)  # 8 batches of 16 targets for the first task
  task2_targets = torch.randn(8, 16, 1)  # 8 batches of 16 targets for the second task
```

      optimizer.zero_grad()
```

### The `autojac` engine

The [`autojac` engine](https://torchjd.org/stable/docs/autojac/) provides fine-grained control
over Jacobian computation and aggregation. It lets you compute Jacobians with respect to specific
layers or activations (partial Jacobian descent), store them in `.jac` fields for inspection, and
apply any aggregator independently. See the [autojac examples](https://torchjd.org/stable/examples/)

> for more details.

Reviewer: "fine-grained control over Jacobian computation and aggregation" -> "Jacobian computation and aggregation utility."

"(partial Jacobian descent)" -> remove

"for inspection" -> "for later usage"?

"apply aggregator independently" -> "apply aggregator flexibly"

Reviewer 2: I also want to add a few things. I'll rewrite this part myself, to make it closer to https://discuss.pytorch.org/t/torchjd-computing-and-aggregating-jacobians-with-pytorch/224655#p-478121-computing-jacobians-2

include clear instructions on how to migrate.

## Contribution

Please read the [Contribution page](CONTRIBUTING.md) and join our

> [![Discord](https://img.shields.io/badge/Discord-%235865F2?logo=discord&logoColor=white)](https://discord.gg/76KkRnb3nk)

Reviewer: I think the badge here looks weird
```suggestion
[Discord](https://discord.gg/76KkRnb3nk)
```
